invokeai/app/services/shared/sqlite/sqlite_migrator.py

```diff
@@ -0,0 +1,210 @@
+            current_version = self._get_current_version()
+
+            if len(sorted_migrations) == 0:
+                self._logger.debug("No migrations registered")
+                return
+
+            latest_version = sorted_migrations[-1].db_version
+            if current_version == latest_version:
+                self._logger.debug("Database is up to date, no migrations to run")
+                return
+
+            if current_version > latest_version:
+                raise MigrationError(
+                    f"Database version {current_version} is greater than the latest migration version {latest_version}"
+                )
+
+            self._logger.info("Database update needed")
+
+            # Only make a backup if using a file database (not memory)
+            backup_path: Optional[Path] = None
+            if isinstance(self._database, Path):
+                backup_path = self._backup_db(self._database)
+            else:
+                self._logger.info("Using in-memory database, skipping backup")
+
+            for migration in sorted_migrations:
+                try:
+                    self._run_migration(migration)
+                except MigrationError:
+                    if backup_path is not None:
+                        self._logger.error(f" Restoring from {backup_path}")
+                        self._restore_db(backup_path)
+                    raise
+            self._logger.info("Database updated successfully")
+
+    def _run_migration(self, migration: Migration) -> None:
+        """Runs a single migration."""
+        with self._lock:
+            current_version = self._get_current_version()
+            try:
+                if current_version >= migration.db_version:
+                    return
+                migration.migrate(self._cursor)
+                # Migration callbacks only get a cursor; they cannot commit the transaction.
+                self._conn.commit()
+                self._set_version(db_version=migration.db_version, app_version=migration.app_version)
+                self._logger.debug(f"Successfully migrated database from {current_version} to {migration.db_version}")
+            except Exception as e:
+                msg = f"Error migrating database from {current_version} to {migration.db_version}: {e}"
+                self._conn.rollback()
+                self._logger.error(msg)
+                raise MigrationError(msg) from e
+
+    def _create_version_table(self) -> None:
+        """Creates a version table for the database, if one does not already exist."""
+        with self._lock:
+            try:
+                self._cursor.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='version';")
+                if self._cursor.fetchone() is not None:
+                    return
+                self._cursor.execute(
+                    """--sql
+                    CREATE TABLE IF NOT EXISTS version (
```

Comment:
nit: `migrations` might better describe what this table contains.

Reply:
Agreed